### ISSUE TITLE
Common extension integration in favor of custom installation

### ISSUE THREAD

**Issue by Jelmerro** [Jelmerro/Vieb#385]
**Checklist**
- [x] I searched the `:help` documentation using `/` for helpful information
- [x] I have read the [frequently asked questions](https://github.com/Jelmerro/Vieb/blob/master/FAQ.md)
- [x] I did not find any similar issues in the [issue tracker](https://github.com/Jelmerro/Vieb/issues)

**Addition or change**

As had been the case for a few years, extension support in Electron isn't good, neither is the support in Vieb. For this reason, I propose to deprecate and remove the current extension support in favor of integrating commonly requested features into Vieb without the need for an extension. The current list of extensions (or features of said extensions) that are considered are:

- [x] Darkreader (implemented and ready for release)
- [x] Sponsorblock (implemented and ready for release)
- [ ] Password manager, see #106 
- [ ] Greasemonkey or similar, see #64 
- [ ] Stylus or similar (to apply custom styling to a page)
- [ ] Scrollbar customization of some sort (could also be part of stylus-like implementation)
- [ ] Scrolling behavior controlled with [scroll-behavior](https://developer.mozilla.org/en-US/docs/Web/CSS/scroll-behavior) and [scroll-snap-type](https://developer.mozilla.org/en-US/docs/Web/CSS/scroll-snap-type)
- [ ] Built-in dictionary (as in, show the definition of a word, not just correct misspelled ones)
- [ ] Word replacer (replace random words on the page with different language words with definition, as to learn languages)
- [ ] Page/word translator (related to the two above)
- [ ] Text-to-speech, read pages or section of it (selected with visual) out loud
- [ ] Automatic clearing of cookies, localstorage and cache, based on a timer and/or domain name
- [ ] Bookmarks #100 

All of these should be OPTIONALLY activated, as is already the case for the ones that are marked as done. Please feel free to suggest additional extensions/features you use often and for which there is no existing way to do so in Vieb, either in the comments below or as a new ticket.

**Alternatives considered**

An alternative to the above scenario would be to create and support browser extensions, but please DO NOT underestimate this task. I have tried and failed for years, as have many others. The enormous size of this task is what led to the deprecation of extension support in the first place. If you think you are up for the task, please go ahead: write the code and make a PR for it, but until then, know that building optional features into Vieb will be the current way forward. For info about the previous extension approach, see #130.

**Comment by yosoymau** [comment]
Something that would be very nice to have related to the word translator and builtiin dictionary would be integration with anki, like the way yomichan on firefox has.

**Comment by Melandel** [comment]
I would definitely enjoy a markdown viewer.

Example on firefox: [markdown-viewer-webext](https://addons.mozilla.org/fr/firefox/addon/markdown-viewer-webext/)

**Comment by saltrafael** [comment]
https://github.com/SimonBrazell/privacy-redirect

**Comment by Jelmerro** [comment]
That seems like it is already possible with the redirects setting, so not adding to the list for now.

**Comment by Nosp1** [comment]
Hi, ive just discovered the browser, and I'm in love! But i need [Modheader plugin ](https://modheader.com/) for day to day development for injecting certain env variables in the http requests. As Ive just discovered this browser i'm not sure if you already support this, but It could be a great feature to integrate (if at all possible).

(I do apologise if my comment here is out of place)

**Comment by Jelmerro** [comment]
This seems like a nice addition, would you need multiple profiles with filters, or would a general method to edit the request headers be enough? And if not, how should we organize the config?

**Comment by Nosp1** [comment]
Thanks for the quick reply and positive onlook. I use about three modheaders at the same time.So perhaps a simple way to toggle them on and off. Here is an example from the extension: 
![image](https://user-images.githubusercontent.com/13243486/229252177-df17d39e-6a2b-40c8-ad90-f8334b108f2c.png).
It might be difficult to implement, I havn't had the time to look at the source code for Vieb yet.
 It may be more feasible with a  config file with simple key-value pairs. Or maybe one could use the command mode and do something in lines of   `set modheader `<key> <value>` and a command for clearing a mod header.
 `set modheader <blank>` to clear all mod headers or `unset modheader <key>` to clear a single one. I think at least being able to have several would be convenient.

**Comment by Nosp1** [comment]
Wow, thats great! Then I can use the browser at work! Awsome!

**Comment by saltrafael** [comment]
suggestion: https://github.com/shevabam/get-rss-feed-url-extension (yank page's RSS feed?)

**Comment by Jelmerro** [comment]
Looked at this briefly, and while it does seem cool and useful to add, the example extension sends every request to the "api url", which is hosted by the owner of the extension, where the logic for extracting the feed urls is abstracted away. This is both a security issue and cannot be used as an example of how to implement this. I would accept contributions to implement client-side feed url extraction, but sending the url to a server to get a list of feed urls seems illogical and overcomplex. Since I have never used RSS in my life for anything, I will not be working on this personally, but I do accept PRs that implement this (again, if parsing of the page is done client-side in Vieb, not sending every url to a server like the extension linked).

**Comment by moneylcj** [comment]
Is it possible to add the ability to access vieb internal resources and execute
commands from runjsinpage, just like vimperator  liberator.execute did. This will open a new
door for users to customize plugins.

**Comment by Jelmerro** [comment]
Not currently, but that is tracked in this ticket https://github.com/Jelmerro/Vieb/issues/192, so please continue discussion for such a feature there.

**Comment by r3k2** [comment]
something like libredirect https://libredirect.github.io/ I know vieb has set redirect but is limited in comparation with this one I point out.. 

ALSO: please gemini/gopher support will be great

**Comment by Jelmerro** [comment]
Redirects in Vieb are less easy to configure, but absolutely more powerful than regular redirect libraries, as you get complete freedom to change any url to a completely different url. While not directly aimed at replacing services with a different (possibly compatible instance) of it, with the right replacement patterns you can achieve the same functionality, but also for custom redirects or configs.

**Comment by Mateiadrielrafael** [comment]
People uwe many external password managers. Is the goal to bake in integration for all of them?

**Comment by Jelmerro** [comment]
Not exactly, but since extension support is pretty much impossible (as per #130), the current approach is to integrate the functionality of extensions for regular feature requests. For password management, there is a separate ticket that should ideally have custom integration with any password manager, see #106.

**Comment by IdoEsports** [comment]
Would a vim client in the browser like in [firenvim](https://github.com/glacambre/firenvim) be possible in Vieb?

I know of the vimcommand setting but I think most people would prefer the above method if it's possible.

**Comment by Jelmerro** [comment]
Sorry for the late reply, but there's a separate issue for that: https://github.com/Jelmerro/Vieb/issues/61. Since firenvim is only for neovim and extensions are not a core part of Electron nor easy to do, I have not worked on this yet, as it would probably need to be build from scratch. The vimcommand option is recommended in the meantime.
